compiler/qsc_parse/src/lex/concrete.rs

```diff
@@ -0,0 +1,132 @@
+            self.non_compilation_tokens =
+                raw::Lexer::new_with_starting_offset(slice, starting_offset).peekable();
+        }
+    }
+
+    fn get_next_lo(&mut self) -> u32 {
+        match self.non_compilation_tokens.peek() {
+            Some(next) => next.offset,
+            None => match self.cooked_tokens.peek() {
+                Some(next) => match next {
+                    Ok(next) => next.span.lo,
+                    Err(err) => err.get_span().lo,
+                },
+                None => self
+                    .code
+                    .len()
+                    .try_into()
+                    .expect("expected length of code to fit into u32"),
+            },
+        }
+    }
+}
+
+impl Iterator for ConcreteTokenIterator<'_> {
+    type Item = ConcreteToken;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.non_compilation_tokens.next() {
+            Some(raw_token) => {
+                let next_lo = self.get_next_lo();
+                let span = Span {
+                    lo: raw_token.offset,
+                    hi: next_lo,
+                };
+                let concrete = match raw_token.kind {
+                    raw::TokenKind::Comment(_) => ConcreteToken {
+                        kind: ConcreteTokenKind::Comment,
+                        span,
+                    },
+                    raw::TokenKind::Whitespace => ConcreteToken {
+                        kind: ConcreteTokenKind::WhiteSpace,
+                        span,
+                    },
+                    raw::TokenKind::Single(raw::Single::Apos) => {
+                        // Apostrophes are handled a bit strangely.
+                        // Their full information is contained in the following token.
+                        return self.next();
+                    }
+                    // This will panic if any content other than whitespace or comments are ignored when "cooking" the raw tokens
+                    _ => panic!("Raw Token couldn't be converted: {raw_token:?}"),
+                };
+                Some(concrete)
+            }
+            None => match self.cooked_tokens.next()? {
+                Ok(token) => {
+                    let next_lo = self.get_next_lo();
+                    self.get_tokens_from_span(token.span.hi, next_lo);
+                    let syntax = ConcreteToken {
+                        kind: ConcreteTokenKind::Syntax(token.kind),
+                        span: token.span,
+                    };
+                    Some(syntax)
+                }
+                Err(err) => {
+                    let next_lo = self.get_next_lo();
+                    let span = err.get_span();
+                    self.get_tokens_from_span(span.hi, next_lo);
+                    let error = ConcreteToken {
+                        kind: ConcreteTokenKind::Error(err),
+                        span,
+                    };
+                    Some(error)
```

Comment:
What exactly happens if we swallow errors and keep going? Is there ever a next token after the error? Should we just stop instead of returning an Error token? An test case would be helpful, e.g. using an unterminated string literal.

Reply:
It should just skip past error tokens. Tests have been added.